### PR TITLE
Add Anne Pro 2 keyboard

### PR DESCRIPTION
**Pull request by bwisn**
## Description

Introduces support for Anne Pro 2

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Review by fauxpark — CHANGES_REQUESTED** [review]
Please submit the core changes as a separate PR.

**Comment by bwisn** [comment]
> Please submit the core changes as a separate PR.
 Removed core and submitted it in another PR:
#14388

**Review by zvecr — CHANGES_REQUESTED** [review]
I would also recommend removing all non-default keymaps from the PR. Theres quite a lot that needs to change (a lot of which is on the PR checklist), which bloats this PR and would be better for them to resubmit after this has merged.

**Comment by tzarc** [comment]
Retargeted to `develop` due to dependent PR requiring core changes.

File: keyboards/annepro2/annepro2.c [hunk 1]
```diff
@@ -0,0 +1,196 @@
+#include "ch.h"
```
Comment (drashna) [inline keyboards/annepro2/annepro2.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/annepro2/c15/config.h [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    K30,      K32, K33, K34, K35, K36, K37, K38, K39, K3A, K3B, K3C,      \
+    K40, K42, K43,                K46,           K49,  K4A,   K4B,   K4C  \
+) { \
+    /*            COL1   COL2   COL3   COL4   COL5   COL6   COL7   COL8   COL9   COL10  COL11  COL12  COL13  COL14*/ \
+    /* ROW1  */ { K00,   K01,   K02,   K03,   K04,   K05,   K06,   K07,   K08,   K09,   K0A,   K0B,   K0C,   K0D  }, \
+    /* ROW2  */ { K10,   K11,   K12,   K13,   K14,   K15,   K16,   K17,   K18,   K19,   K1A,   K1B,   K1C,   K1D  }, \
+    /* ROW3  */ { K20,   K21,   K22,   K23,   K24,   K25,   K26,   K27,   K28,   K29,   K2A,   K2B,   K2C,   KC_NO}, \
+    /* ROW4  */ { K30,   KC_NO, K32,   K33,   K34,   K35,   K36,   K37,   K38,   K39,   K3A,   K3B,   K3C,   KC_NO}, \
+    /* ROW5  */ { K40,   KC_NO, K42,   K43,   KC_NO, KC_NO, K46,   KC_NO, KC_NO, K49,   K4A,   K4B,   K4C,   KC_NO}, \
+}
```
Comment (drashna) [inline keyboards/annepro2/c15/config.h:57]: This should go in the keyboard's header file, not here. 
```suggestion
```

File: keyboards/annepro2/c15/config.h [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+//#define MAGIC_KEY_LAYER5         5
+//#define MAGIC_KEY_LAYER6         6
+//#define MAGIC_KEY_LAYER7         7
+//#define MAGIC_KEY_LAYER8         8
+//#define MAGIC_KEY_LAYER9         9
+//#define MAGIC_KEY_BOOTLOADER     R
+//#define MAGIC_KEY_LOCK           CAPS
+//#define MAGIC_KEY_EEPROM         E
+//#define MAGIC_KEY_NKRO           N
+//#define MAGIC_KEY_SLEEP_LED      Z
```
Comment (drashna) [inline keyboards/annepro2/c15/config.h:149]:
```suggestion
```

File: keyboards/annepro2/c15/rules.mk [hunk 4]
```diff
@@ -0,0 +1,41 @@
+# Anne Pro 2
+SRC = \
+	matrix.c \
+	hardfault_handler.c \
+	annepro2_ble.c \
+	qmk_ap2_led.c \
+	protocol.c
+
+LAYOUTS +=
```
Comment (drashna) [inline keyboards/annepro2/c15/rules.mk:9]:
```suggestion
```

File: keyboards/annepro2/c15/rules.mk [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+BOARD = ANNEPRO2
+
+OPT_DEFS = -Wno-unused-function -fdump-rtl-dfinish -fstack-usage
+#EXTRALDFLAGS = -Wl,--print-memory-usage
+
+# Options
+
+# Keys
+CUSTOM_MATRIX = yes
```
Comment (drashna) [inline keyboards/annepro2/c15/rules.mk:28]: This should be converted to matrix lite, if/when possible. 

https://docs.qmk.fm/#/custom_matrix?id=lite

File: keyboards/annepro2/c18/config.h [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+//#define MAGIC_KEY_LAYER5         5
+//#define MAGIC_KEY_LAYER6         6
+//#define MAGIC_KEY_LAYER7         7
+//#define MAGIC_KEY_LAYER8         8
+//#define MAGIC_KEY_LAYER9         9
+//#define MAGIC_KEY_BOOTLOADER     R
+//#define MAGIC_KEY_LOCK           CAPS
+//#define MAGIC_KEY_EEPROM         E
+//#define MAGIC_KEY_NKRO           N
+//#define MAGIC_KEY_SLEEP_LED      Z
```
Comment (drashna) [inline keyboards/annepro2/c18/config.h:174]:
```suggestion
```

File: keyboards/annepro2/c18/rules.mk [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+BOARD = ANNEPRO2_C18
+
+OPT_DEFS = -Wno-unused-function -fdump-rtl-dfinish -fstack-usage
+#EXTRALDFLAGS = -Wl,--print-memory-usage
+
+# Options
+
+# Keys
+CUSTOM_MATRIX = yes
```
Comment (drashna) [inline keyboards/annepro2/c18/rules.mk:33]: Same as above (matrix lite)

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by bwisn** [comment]
@drashna fixed!

File: keyboards/annepro2/info.json [hunk 8]
```diff
@@ -0,0 +1,332 @@
+{
+    "keyboard_name": "Anne Pro 2",
+    "url": "https://openannepro.github.io/",
+    "maintainer": "community",
+    "width": 15,
+    "height": 5,
```
Comment (drashna) [inline keyboards/annepro2/info.json:6]:
```suggestion
```

File: keyboards/annepro2/keymaps/default-full-caps/keymap.c [hunk 9]
```diff
@@ -0,0 +1,127 @@
+#include <stdint.h>
+#include "annepro2.h"
+#include "qmk_ap2_led.h"
+#include "config.h"
```
Comment (drashna) [inline keyboards/annepro2/keymaps/default-full-caps/keymap.c:4]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

And:
```suggestion
#include QMK_KEYBOARD_H
```

File: keyboards/annepro2/keymaps/default-full-caps/config.h [hunk 10]
```diff
@@ -0,0 +1,2 @@
+// Obins stock firmware has something similar to this already enabled, but disabled by default in QMK
```
Comment (drashna) [inline keyboards/annepro2/keymaps/default-full-caps/config.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

```suggestion
#pragma once
// Obins stock firmware has something similar to this already enabled, but disabled by default in QMK
```

File: keyboards/annepro2/keymaps/default-full-caps/keymap.c [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+    // annepro2-shine disables LEDs by default. Uncomment this function to enable them at startup.
+    // annepro2LedEnable();
+
+    // Additionally, it also chooses the first LED profile by default. Refer to the "profiles" array in main.c in
+    // annepro2-shine to see the order. Replace "i" with the index of your preferred profile. (i.e the RED profile is index 0)
+    // annepro2LedSetProfile(i);
+}
+
+layer_state_t layer_state_set_user(layer_state_t layer) { return layer; }
```
Comment (drashna) [inline keyboards/annepro2/keymaps/default-full-caps/keymap.c:115]:
```suggestion

// Code to run after initializing the keyboard
void keyboard_post_init_user(void) {
    // Here are two common functions that you can use. For more LED functions, refer to the file "qmk_ap2_led.h"

    // annepro2-shine disables LEDs by default. Uncomment this function to enable them at startup.
    // annepro2LedEnable();

    // Additionally, it also chooses the first LED profile by default. Refer to the "profiles" array in main.c in
    // annepro2-shine to see the order. Replace "i" with the index of your preferred profile. (i.e the RED profile is index 0)
    // annepro2LedSetProfile(i);
}
```

File: keyboards/annepro2/keymaps/default-layer-indicators/config.h [hunk 12]
```diff
@@ -0,0 +1,2 @@
+// Obins stock firmware has something similar to this already enabled, but disabled by default in QMK
+#define PERMISSIVE_HOLD
```
Comment (drashna) [inline keyboards/annepro2/keymaps/default-layer-indicators/config.h:2]: Same as previous

File: keyboards/annepro2/keymaps/default-layer-indicators/keymap.c [hunk 13]
```diff
@@ -0,0 +1,151 @@
+#include <stdint.h>
```
Comment (drashna) [inline keyboards/annepro2/keymaps/default-layer-indicators/keymap.c:1]: Same as previous

File: keyboards/annepro2/keymaps/default-layer-indicators/keymap.c [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+    // annepro2-shine disables LEDs by default. Uncomment this function to enable them at startup.
+    // annepro2LedEnable();
+
+    // Additionally, it also chooses the first LED profile by default. Refer to the "profiles" array in main.c in
+    // annepro2-shine to see the order. Replace "i" with the index of your preferred profile. (i.e the RED profile is index 0)
+    // annepro2LedSetProfile(i);
+}
+
+layer_state_t layer_state_set_user(layer_state_t layer) {
+    switch (get_highest_layer(layer)) {
```
Comment (drashna) [inline keyboards/annepro2/keymaps/default-layer-indicators/keymap.c:115]: While the name doesn't technically matter to the compiler, this is a bitmask of the layer state, so `state` is a bit more accurate and clear. 
```suggestion
layer_state_t layer_state_set_user(layer_state_t state) {
    switch (get_highest_layer(state)) {
```

File: keyboards/annepro2/keymaps/default/config.h [hunk 15]
```diff
@@ -0,0 +1,2 @@
+// Obins stock firmware has something similar to this already enabled, but disabled by default in QMK
```
Comment (drashna) [inline keyboards/annepro2/keymaps/default/config.h:1]: Same as previous

File: keyboards/annepro2/keymaps/default/keymap.c [hunk 16]
```diff
@@ -0,0 +1,112 @@
+#include <stdint.h>
```
Comment (drashna) [inline keyboards/annepro2/keymaps/default/keymap.c:1]: same as previous

File: keyboards/annepro2/qmk_ap2_led.c [hunk 17]
```diff
@@ -0,0 +1,118 @@
+#include <string.h>
```
Comment (drashna) [inline keyboards/annepro2/qmk_ap2_led.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/annepro2/qmk_ap2_led.h [hunk 18]
```diff
@@ -0,0 +1,68 @@
+#pragma once
```
Comment (drashna) [inline keyboards/annepro2/qmk_ap2_led.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/annepro2/annepro2.h [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+#include "quantum.h"
+#include <stdint.h>
+#include "qmk_ap2_led.h"
+#include "config.h"
```
Comment (drashna) [inline keyboards/annepro2/annepro2.h:22]: This line is not needed, at all. 
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/annepro2/keymaps/default-full-caps/keymap.c [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+* |-----------------------------------------------------------------------------------------+
+* |         |     |     |     |     |     |     |     |     |     |     |     |             |
+* |-----------------------------------------------------------------------------------------+
+* |            |     |     |     |     |     |     |     |     |     |     |       UP       |
+* |-----------------------------------------------------------------------------------------+
+* |       |       |       |                                 |       |  LEFT | DOWN  | RIGHT |
+* \-----------------------------------------------------------------------------------------/
+*/
+ const uint16_t keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+ [_BASE_LAYER] = KEYMAP( /* Base */
```
Comment (fauxpark) [inline keyboards/annepro2/keymaps/default-full-caps/keymap.c:55]: These need to be updated.

**Comment by rpeshkov** [comment]
I have the same problem as for @tzarc. Keyboard gets flashed, but it's not functional. Previously I had success with repo https://github.com/OpenAnnePro/qmk_firmware.

Played around with the code in this PR. For me reverting back `CUSTOM_MATRIX = yes` and restoring matrix methods did the trick and my Anne Pro started to work. I have C18 version.

**Comment by tzarc** [comment]
For reference, I've left this PR alone as you've resolved comments on the GitHub UI, but not fixed them in the code.
We spoke on Discord about this, you apologised, and they're still unaddressed.

I'd suggest going through each and every comment made by myself, @drashna and @zvecr, ensuring every item has been addressed.

Feel free to ping here when done -- I'll go through every request to ensure it's been handled.

File: keyboards/annepro2/annepro2_ble.c [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+        default:
+            return 0x00;
+    }
+}
+
+static void ap2_ble_consumer(uint16_t data) {
+    sdPut(&SD1, 0x0);
+    sdWrite(&SD1, bleMcuSendConsumerReport, 10);
+    sdPut(&SD1, CONSUMER2AP2(data));
+    sdWrite(&SD1, 0x0, 3);
```
Comment (tzarc) [inline keyboards/annepro2/annepro2_ble.c:159]:
```suggestion
    static const uint8_t dummy[3] = {0};
    sdWrite(&SD1, dummy, sizeof(dummy));
```

File: keyboards/annepro2/boards/ANNEPRO2_C15/board.c [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+#include <string.h>
+#include "hal.h"
+#include "wait.h"
+#include "led.h"
+#include "annepro2.h"
```
Comment (tzarc) [inline keyboards/annepro2/boards/ANNEPRO2_C15/board.c:21]: ChibiOS board definitions should not include keyboard files.
```suggestion
#include "hal.h"
```

File: keyboards/annepro2/boards/ANNEPRO2_C15/board.c [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+    LINE_COL1, LINE_COL2, LINE_COL3, LINE_COL4, LINE_COL5, LINE_COL6, LINE_COL7, LINE_COL8, LINE_COL9, LINE_COL10, LINE_COL11, LINE_COL12, LINE_COL13, LINE_COL14,
+};
+
+void __early_init(void) {}
+
+void early_hardware_init_pre(void) { ht32_clock_init(); }
+
+void board_init(void) {}
+
+void boardInit(void) {}
```
Comment (tzarc) [inline keyboards/annepro2/boards/ANNEPRO2_C15/board.c:119]: These are QMK-specific constructs and do not belong in the `board.c`.
You've misinterpreted the intent here -- `ht32_clock_init();` does indeed go into `__early_init()`, as per normal board files for stm32, which place `stm32_clock_init()` in the same location.

As I pointed out on specific lines highlighted in the previous comment, the issue was the `gpt_init()` and the `spi_init()`, not the placement of the clock init.

Refer to an existing ChibiOS board file for example structure.

```suggestion
void __early_init(void) { ht32_clock_init(); }

void boardInit(void) {}
```

File: keyboards/annepro2/boards/ANNEPRO2_C15/board.h [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+#define LINE_COL6 PAL_LINE(IOPORTA, 15)
+#define LINE_COL7 PAL_LINE(IOPORTA, 8)
+#define LINE_COL8 PAL_LINE(IOPORTA, 10)
+#define LINE_COL9 PAL_LINE(IOPORTA, 11)
+// PORTA 12,13 conflict with SWD
+#define LINE_COL10 PAL_LINE(IOPORTA, 12)
+#define LINE_COL11 PAL_LINE(IOPORTA, 13)
+#define LINE_COL12 PAL_LINE(IOPORTA, 14)
+#define LINE_COL13 PAL_LINE(IOPORTB, 2)
+#define LINE_COL14 PAL_LINE(IOPORTB, 3)
```
Comment (tzarc) [inline keyboards/annepro2/boards/ANNEPRO2_C15/board.h:64]: These are keyboard-specific, and should be swapped out for normal GPIO definitions as per the MCU's datasheet.

The config.h file for each board can then be keyboard-specific and map the rows/cols to the MCU pin definitions.

```suggestion
```

File: keyboards/annepro2/boards/ANNEPRO2_C15/board.h [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+
+#pragma once
+/*
+ * Setup for Anne Pro 2 board.
+ */
+
+/*
+ * Board identifier.
+ */
+#define ANNEPRO2
```
Comment (tzarc) [inline keyboards/annepro2/boards/ANNEPRO2_C15/board.h:25]:
```suggestion
```

File: keyboards/annepro2/boards/ANNEPRO2_C18/board.h [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+
+#pragma once
+/*
+ * Setup for Anne Pro 2 board.
+ */
+
+/*
+ * Board identifier.
+ */
+#define ANNEPRO2
```
Comment (tzarc) [inline keyboards/annepro2/boards/ANNEPRO2_C18/board.h:25]:
```suggestion
```

File: keyboards/annepro2/boards/ANNEPRO2_C18/board.h [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+#define LINE_COL6                   PAL_LINE(IOPORTA, 15)
+#define LINE_COL7                   PAL_LINE(IOPORTC, 12)
+#define LINE_COL8                   PAL_LINE(IOPORTC, 13)
+#define LINE_COL9                   PAL_LINE(IOPORTA,  8)
+#define LINE_COL10                  PAL_LINE(IOPORTA, 10)
+#define LINE_COL11                  PAL_LINE(IOPORTA, 11)
+#define LINE_COL12                  PAL_LINE(IOPORTA, 14)
+#define LINE_COL13                  PAL_LINE(IOPORTD,  2)
+#define LINE_COL14                  PAL_LINE(IOPORTD,  3)
+
```
Comment (tzarc) [inline keyboards/annepro2/boards/ANNEPRO2_C18/board.h:63]: As per C15.
```suggestion
```

File: keyboards/annepro2/c15/rules.mk [hunk 28]
```diff
@@ -0,0 +1,39 @@
+# Anne Pro 2
+SRC = \
+	matrix.c \
+	hardfault_handler.c \
```
Comment (tzarc) [inline keyboards/annepro2/c15/rules.mk:4]:
```suggestion
```

File: keyboards/annepro2/c18/rules.mk [hunk 29]
```diff
@@ -0,0 +1,39 @@
+# Anne Pro 2
+SRC = \
+	matrix.c \
+	hardfault_handler.c \
```
Comment (tzarc) [inline keyboards/annepro2/c18/rules.mk:4]:
```suggestion
```

File: keyboards/annepro2/chconf.h [hunk 30]
```diff
@@ -0,0 +1,70 @@
...
+
+#define CH_CFG_FACTORY_OBJ_FIFOS TRUE
+
+#define CH_CFG_FACTORY_PIPES TRUE
+
+#define CH_DBG_SYSTEM_STATE_CHECK FALSE  // adds about 1K code
+
+#define CH_DBG_ENABLE_CHECKS FALSE  // adds < 1K code
+
+#define CH_DBG_ENABLE_ASSERTS FALSE  // adds about 3K code
```
Comment (tzarc) [inline keyboards/annepro2/chconf.h:68]: Remove as much of these as is possible.
This looks like copypaste without understanding the implications.

```suggestion
```

File: keyboards/annepro2/mcuconf.h [hunk 31]
```diff
@@ -0,0 +1,69 @@
...
+
+#define HT32_USB_USE_USB0 TRUE
+#define HT32_USB_USB0_IRQ_PRIORITY 5
+
+#if defined(ANNEPRO2_EEPROM)
+#define HT32_SPI_USE_SPI1 TRUE
+#define HT32_SPI1_IRQ_PRIORITY 9
+#endif
+
+
```
Comment (tzarc) [inline keyboards/annepro2/mcuconf.h:69]:
```suggestion
```

File: keyboards/annepro2/protocol.h [hunk 32]
```diff
@@ -0,0 +1,111 @@
+/*
+ * (c) 2021 by Tomasz bla Fortuna
+ * License: GPLv2
+ *
+ * This file is shared with the Shine firmware. Keep it in sync (and in the
+ * shine's clang formatting).
+ */
+
+#pragma once
+#include <inttypes.h>
```
Comment (tzarc) [inline keyboards/annepro2/protocol.h:10]:
```suggestion
#include <stdint.h>
```

File: keyboards/annepro2/readme.md [hunk 33]
```diff
@@ -0,0 +1,27 @@
...
+
+    make annepro2/c18:default
+
+If you want to compile the Anne Pro 2 C15 default keymap use:
+
+    make annepro2/c15:default
+
+## Installing
+
+See the [installing instructions](https://openannepro.github.io/install/)
```
Comment (tzarc) [inline keyboards/annepro2/readme.md:25]: This document needs flashing instructions embedded here, rather than linked to.

```suggestion
```

**Review by tzarc — CHANGES_REQUESTED** [review]
Round 2.

**Comment by bwisn** [comment]
@tzarc friendly ping :) IMO everything is fixed, waiting for review

**Comment by blaa** [comment]
We can wait, but maybe there's something more we can do. We've got a recent branch with a better RGB integration with the QMK (letting QMK control LEDs via commands to Shine/LED CPU), it still needs tests on C15 variant (less RAM) though. Our original idea was to leave it as it is, well-tested and work on bells and whistles later.

**Review by drashna — APPROVED** [review]

**Comment by bwisn** [comment]
Please change target from master to develop. It needs #16548 to be merged first.

**Review by tzarc — APPROVED** [review]
Close enough.

**Pull request merged**